aeron-cluster/src/main/java/io/aeron/cluster/Election.java

```diff
@@ -707,21 +714,49 @@ private int followerLogReplication(final long nowNs)
         else
         {
             workCount += consensusModuleAgent.pollArchiveEvents();
-            if (logReplication.isDone())
+            if (logReplication.isDone(nowNs))
             {
-                appendPosition = logReplication.position();
-                consensusModuleAgent.logRecordingId(logReplication.recordingId());
-
-                if (consensusPublisher.appendPosition(
-                    leaderMember.publication(), leadershipTermId, appendPosition, thisMember.id()))
+                if (logReplication.isClosed())
                 {
-                    // TODO: only begin replay when commit position from leader is received that matches
-                    cleanupReplication();
-                    state(FOLLOWER_REPLAY, nowNs);
+                    if (leaderLogReplicationCommitPosition >= appendPosition)
+                    {
+                        cleanupReplication();
+                        state(FOLLOWER_REPLAY, nowNs);
+
+                        workCount++;
+                    }
+                    else if (nowNs >= leaderLogReplicationCommitPositionDeadlineNs)
+                    {
+                        throw new ClusterException("Timeout waiting for leader to commit replicated log position");
+                    }
+                    else
+                    {
+                        if (nowNs + ctx.leaderHeartbeatIntervalNs() >= timeOfLastUpdateNs)
+                        {
+                            if (consensusPublisher.appendPosition(
+                                leaderMember.publication(), leadershipTermId, appendPosition, thisMember.id()))
+                            {
+                                timeOfLastUpdateNs = nowNs;
+                            }
+
+                            workCount++;
+                        }
+                    }
                 }
                 else
                 {
-                    // TODO: check for timeout of leader liveness
+                    logReplication.close();
+                    appendPosition = logReplication.position();
+                    consensusModuleAgent.logRecordingId(logReplication.recordingId());
+                    if (consensusPublisher.appendPosition(
+                        leaderMember.publication(), leadershipTermId, appendPosition, thisMember.id()))
+                    {
+                        timeOfLastUpdateNs = nowNs;
+                    }
+
+                    leaderLogReplicationCommitPositionDeadlineNs = nowNs + ctx.leaderHeartbeatTimeoutNs();
```

Comment:
Should this be updated here?

Reply:
I've changed this around a bit now.  Previously I was relying on the transition to closed on the `LogReplication` to manage a lot of these updates.  Instead now the LogReplication is either done or not and the logic is similar in both cases.  It will also pulse out the append position as the replication progresses (if it takes longer than the leader heartbeat interval).  I've updating the `leaderLogReplicationCommitPositionDeadlineNs` whenever the append position moves forward (so eventually once log replication is done, will it have the one `leaderHeartbeatTimeoutNs` period in which to commit the final append position).